File: src/engine/staff.ts
/**
 * Coaching staff model (spec §16) — a small set of quality ratings plus explicit
 * tendency knobs for the head coach and the two coordinators.
 *
 * These feed `staff-shift.ts`, which nudges the same resolver logits/means the
 * player rating layer touches. Effects are deliberately subtle: a best-vs-worst
 * full staff is worth roughly 3–4 points per team-game, so individual games
 * barely move and the §22 validation is unaffected. When no staff is supplied
 * the shift is exactly zero (the rating-parity path).
 *
 * v0 values are authored (`staff-data.ts`) on a compressed 35–70 scale and are
 * expected to be tuned.
 */

export type OffScheme =
  | "west_coast"
  | "vertical"
  | "spread"
  | "power_run"
  | "zone_run"
  | "pro_style";

export type DefScheme =
  | "four_three"
  | "three_four"
  | "multiple"
  | "cover_3"
  | "cover_2"
  | "man_press";

export interface HeadCoach {
  name: string;
  /** 1–99. Situational decisions: 4th downs, timeouts, two-point/challenge calls. */
  gameManagement: number;
  /** 1–99. Team discipline → fewer pre-snap and live-ball penalties. */
  discipline: number;
  /** −1..+1 relative to the league-average 4th-down go rate. */
  aggression: number;
}

export interface OffensiveCoordinator {
  name: string;
  /** 1–99. Scheme design + in-game adjustments → offensive efficiency. */
  rating: number;
  scheme: OffScheme;
  /** −1..+1 pass rate over expectation (texture; v0.1 wires this to M02). */
  passBias: number;
  /** −1..+1 snap tempo (positive = faster / more no-huddle). */
  tempo: number;
}

export interface DefensiveCoordinator {
  name: string;
  /** 1–99. Scheme design + in-game adjustments → efficiency allowed. */
  rating: number;
  scheme: DefScheme;
  /** −1..+1 blitz rate over the league average. */
  blitzBias: number;
}

export interface Staff {
  headCoach: HeadCoach;
  oc: OffensiveCoordinator;
  dc: DefensiveCoordinator;
}

/** A perfectly neutral staff — every shift it produces is zero. */
export function leagueAverageStaff(label = "League Average"): Staff {
  return {
    headCoach: { name: `${label} HC`, gameManagement: 50, discipline: 50, aggression: 0 },
    oc: { name: `${label} OC`, rating: 50, scheme: "pro_style", passBias: 0, tempo: 0 },
    dc: { name: `${label} DC`, rating: 50, scheme: "multiple", blitzBias: 0 },
  };
}

/** Map a 1–99 rating to roughly [-1, +1] (50 → 0), clipped. */
export function ratingNorm(rating: number): number {
  return Math.max(-1, Math.min(1, (rating - 50) / 40));
}

/** scheme_tag values that a coordinator's scheme is a good fit for. */
export const OFF_SCHEME_TAGS: Record<OffScheme, readonly string[]> = {
  west_coast: ["west_coast", "play_action", "move_te", "zone_run", "outside_zone"],
  vertical: ["vertical", "spread", "play_action", "downhill"],
  spread: ["spread", "rpo", "zone_run", "outside_zone", "west_coast"],
  power_run: ["power_run", "gap_scheme", "inline", "downhill", "pass_pro"],
  zone_run: ["zone_run", "outside_zone", "west_coast", "move_te"],
  pro_style: ["play_action", "inline", "move_te", "power_run", "west_coast"],
};

export const DEF_SCHEME_TAGS: Record<DefScheme, readonly string[]> = {
  four_three: ["base_4_3", "one_gap", "penetrate", "attack", "wide_9"],
  three_four: ["base_3_4", "two_gap", "contain", "nose"],
  multiple: ["nickel", "cover_3", "split_safety", "robber", "move_te"],
  cover_3: ["cover_3", "single_high", "zone", "robber"],
  cover_2: ["cover_2", "split_safety", "zone"],
  man_press: ["man_press", "cover_man", "cover_1", "nickel"],
};

/**
 * Fraction of `tagLists` (one per player in a unit) that overlap `schemeTags`.
 * The league-neutral baseline is ~0.4; `staff-shift` centres on that.
 */
export function schemeFitFraction(
  tagLists: readonly (readonly string[] | undefined)[],
  schemeTags: readonly string[],
): number {
  const want = new Set(schemeTags);
  let fit = 0;
  let n = 0;
  for (const tags of tagLists) {
    if (!tags || tags.length === 0) continue;
    n += 1;
    if (tags.some((t) => want.has(t))) fit += 1;
  }
  return n ? fit / n : 0;
}

File: analysis/engine/staff.py
"""Coaching staff model (spec §16) — Python mirror of ``src/engine/staff.ts``.

A small set of quality ratings plus explicit tendency knobs for the head coach
and the two coordinators. Feeds ``staff_shift.py``, which nudges the same
resolver logits/means the player rating layer touches. Effects are deliberately
subtle (best-vs-worst full staff ~= 3-4 points/team-game). No staff -> zero shift.

Kept behaviourally identical to the TypeScript module.
"""

from __future__ import annotations

from dataclasses import dataclass

OFF_SCHEMES = (
    "west_coast", "vertical", "spread", "power_run", "zone_run", "pro_style",
)
DEF_SCHEMES = (
    "four_three", "three_four", "multiple", "cover_3", "cover_2", "man_press",
)


@dataclass(frozen=True)
class HeadCoach:
    name: str
    game_management: float  # 1-99: 4th downs, timeouts, situational calls
    discipline: float       # 1-99: team discipline -> fewer penalties
    aggression: float       # -1..+1 vs the league-average 4th-down go rate


@dataclass(frozen=True)
class Coordinator:
    name: str
    rating: float           # 1-99: scheme design + in-game adjustments
    scheme: str
    tendency: float = 0.0   # OC: pass rate over expectation. DC: blitz rate. (-1..+1)
    tempo: float = 0.0      # OC only: -1..+1 snap tempo (positive = faster)


@dataclass(frozen=True)
class Staff:
    head_coach: HeadCoach
    oc: Coordinator
    dc: Coordinator


def league_average_staff(label: str = "League Average") -> Staff:
    """A perfectly neutral staff — every shift it produces is zero."""
    return Staff(
        head_coach=HeadCoach(f"{label} HC", 50.0, 50.0, 0.0),
        oc=Coordinator(f"{label} OC", 50.0, "pro_style", 0.0, 0.0),
        dc=Coordinator(f"{label} DC", 50.0, "multiple", 0.0, 0.0),
    )


def rating_norm(rating: float) -> float:
    """Map a 1-99 rating to roughly [-1, +1] (50 -> 0), clipped."""
    return max(-1.0, min(1.0, (rating - 50.0) / 40.0))


# scheme_tag values each scheme is a good fit for (mirror of the TS maps).
OFF_SCHEME_TAGS = {
    "west_coast": ("west_coast", "play_action", "move_te", "zone_run", "outside_zone"),
    "vertical": ("vertical", "spread", "play_action", "downhill"),
    "spread": ("spread", "rpo", "zone_run", "outside_zone", "west_coast"),
    "power_run": ("power_run", "gap_scheme", "inline", "downhill", "pass_pro"),
    "zone_run": ("zone_run", "outside_zone", "west_coast", "move_te"),
    "pro_style": ("play_action", "inline", "move_te", "power_run", "west_coast"),
}
DEF_SCHEME_TAGS = {
    "four_three": ("base_4_3", "one_gap", "penetrate", "attack", "wide_9"),
    "three_four": ("base_3_4", "two_gap", "contain", "nose"),
    "multiple": ("nickel", "cover_3", "split_safety", "robber", "move_te"),
    "cover_3": ("cover_3", "single_high", "zone", "robber"),
    "cover_2": ("cover_2", "split_safety", "zone"),
    "man_press": ("man_press", "cover_man", "cover_1", "nickel"),
}


def scheme_fit_fraction(tag_lists, scheme_tags) -> float:
    """Fraction of `tag_lists` (one per player) that overlap `scheme_tags`."""
    want = set(scheme_tags)
    fit = n = 0
    for tags in tag_lists:
        if not tags:
            continue
        n += 1
        if any(t in want for t in tags):
            fit += 1
    return fit / n if n else 0.0
